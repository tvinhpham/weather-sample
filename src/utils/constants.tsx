//export const API_URL = 'https://www.metaweather.com';

//export const API_URL = 'https://cors-anywhere.herokuapp.com/https://www.metaweather.com';
export const API_URL = 'https://meta-weather.vercel.app';
