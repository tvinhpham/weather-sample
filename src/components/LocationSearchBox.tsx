import React from 'react';
import axios from 'axios';
import { FunctionComponent, HTMLAttributes, useState } from 'react';
import { AsyncTypeahead } from 'react-bootstrap-typeahead';
import { useQuery } from 'react-query';

import 'react-bootstrap-typeahead/css/Typeahead.min.css';
import InputGroup from 'react-bootstrap/InputGroup';
import Alert from 'react-bootstrap/Alert';
import { API_URL } from '../utils/constants';

const searchLocations = async ({ queryKey }: any) => {
  let [, query] = queryKey;

  if (!query) {
    return [];
  }

  const { data } = await axios.get(`${API_URL}/api/location/search/?query=${query}`);
  return data;
};

const LocationSearchBox: FunctionComponent<
  {
    onLocationSelected?: (loc: any) => void;
  } & HTMLAttributes<HTMLElement>
> = ({ onLocationSelected, ...props }) => {
  const [searchText, setSearchText] = useState('');

  const { isLoading, isError, data } = useQuery(['locations', searchText], searchLocations);

  const onSearch = (query: string) => {
    setSearchText(query);
  };

  return (
    <div {...props}>
      <InputGroup {...props}>
        <InputGroup.Text>
          <img
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8
            /9hAAAAAXNSR0IArs4c6QAAAQ1JREFUOE+d078rhlEYxvHPSwabMluVgfJHGAw2WcQkpchk4FWKjUFJJJQSSdnY/QdKySL
            /gcIiKXTred4eT8+Pek+d5Vx33/u6r3NOQ/nqwHeF/ic1cgW9WEIfvtCFHxzhtgiWBfRjCyt4yBR3o4l3bOYhKSA6XWEar
            yW2V3GHm6yeAiYTy5cVM0eTM0wUAWLGBXzUhLaHZbyldamDU0zVJY51nOA5D9jFGl5qIMeYzzpNHYxgADsVgJ5Ej6BbK3uN
            59jAYwEkHlV038Z9GSA6HOIaF/hMCgeT8YYSwH4ZIM6j0zhG0ZnsJ0T6YzjAIiKzwqdcdxEzeUj+L9QBQv8HaQcQkDkMY7Z
            dQMvpL1evLxHmgIo1AAAAAElFTkSuQmCC"
            alt="Search icon"
          />
        </InputGroup.Text>

        <AsyncTypeahead
          placeholder="Search"
          data-testid="searchLocationBox"
          // @ts-ignore
          labelKey="title"
          filterBy={() => true}
          id="locationBox"
          options={data || []}
          onChange={([loc]) => onLocationSelected?.(loc)}
          minLength={2}
          renderMenuItemChildren={(option: any) => (
            <div key={option.woeid}>
              <span>{option.title}</span>
            </div>
          )}
          onSearch={onSearch}
          isLoading={isLoading}
        />
      </InputGroup>
      {isError && (
        <Alert key="locationSearchError" variant="danger">
          Unexpected error occurs: Cannot connect to server
        </Alert>
      )}
    </div>
  );
};

export default LocationSearchBox;
