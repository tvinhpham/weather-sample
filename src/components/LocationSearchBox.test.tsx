import React from 'react';
import renderer from 'react-test-renderer';
import LocationSearchBox from './LocationSearchBox';
import { queryClient } from '../utils/helper';
import { QueryClientProvider } from 'react-query';

it('renders search box', () => {

  // @ts-ignore
  const searchBox = renderer
    .create(<QueryClientProvider client={queryClient}><LocationSearchBox /></QueryClientProvider>)
    .toJSON();
  expect(searchBox).toMatchSnapshot();
});
