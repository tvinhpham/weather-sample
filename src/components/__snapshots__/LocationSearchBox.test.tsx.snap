// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders search box 1`] = `
<div>
  <div
    className="input-group"
  >
    <span
      className="input-group-text"
    >
      <img
        alt="Search icon"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8
 /9hAAAAAXNSR0IArs4c6QAAAQ1JREFUOE+d078rhlEYxvHPSwabMluVgfJHGAw2WcQkpchk4FWKjUFJJJQSSdnY/QdKySL
 /gcIiKXTred4eT8+Pek+d5Vx33/u6r3NOQ/nqwHeF/ic1cgW9WEIfvtCFHxzhtgiWBfRjCyt4yBR3o4l3bOYhKSA6XWEar
 yW2V3GHm6yeAiYTy5cVM0eTM0wUAWLGBXzUhLaHZbyldamDU0zVJY51nOA5D9jFGl5qIMeYzzpNHYxgADsVgJ5Ej6BbK3uN
 59jAYwEkHlV038Z9GSA6HOIaF/hMCgeT8YYSwH4ZIM6j0zhG0ZnsJ0T6YzjAIiKzwqdcdxEzeUj+L9QBQv8HaQcQkDkMY7Z
 dQMvpL1evLxHmgIo1AAAAAElFTkSuQmCC"
      />
    </span>
    <div
      className="rbt has-aux"
      style={
        Object {
          "outline": "none",
          "position": "relative",
        }
      }
      tabIndex={-1}
    >
      <div
        style={
          Object {
            "display": "flex",
            "flex": 1,
            "height": "100%",
            "position": "relative",
          }
        }
      >
        <input
          aria-autocomplete="both"
          aria-expanded={false}
          aria-haspopup="listbox"
          autoComplete="off"
          className="rbt-input-main form-control rbt-input"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Search"
          role="combobox"
          type="text"
          value=""
        />
        <input
          aria-hidden={true}
          className="rbt-input-hint"
          readOnly={true}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "boxShadow": "none",
              "color": "rgba(0, 0, 0, 0.54)",
              "left": 0,
              "pointerEvents": "none",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
          tabIndex={-1}
          value=""
        />
      </div>
      <div
        className="rbt-aux"
      >
        <div
          className="rbt-loader spinner-border spinner-border-sm"
          role="status"
        >
          <span
            className="sr-only visually-hidden"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
